File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
from collections import deque
input=sys.stdin.readline

parent={}
child={}
n=int(input())
for i in range(1,n+1):
  parent[i]=0
  child[i]=[]

a,b=map(int,input().split())
m=int(input())
for i in range(m):
  x,y=map(int,input().split())
  child[x].append(y)
  parent[y]=x

visit=[False for i in range(n+1)]
visit[b]=True
q=deque([[b,0]])
while q:
  item,rel=q.popleft()
  if item==a:
    print(rel)
    exit(0)
  if parent[item]!=0 :
    if not visit[parent[item]]:
      visit[parent[item]]=True
      q.append([parent[item],rel+1])
    for nextItem in child[parent[item]]:
      if not visit[nextItem]:
        visit[nextItem]=True
        q.append([nextItem,rel+2])
  for nextItem in child[item]:
    if not visit[nextItem]:
      visit[nextItem]=True
      q.append([nextItem,rel+1])

print(-1)

File: 백준/Silver/2644. 촌수계산/촌수계산.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
let iter = 0;
rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  function find(idx, num) {
    if (parent[idx] !== 0 && !visit[parent[idx]]) {
      if (parent[idx] === b) {
        console.log(num + 1);
        process.exit();
      }
      visit[parent[idx]] = true;
      find(parent[idx], num + 1);
    }
    const childList = Object.keys(child[idx]);
    for (let i = 0; i < childList.length; i++) {
      const mychild = Number(childList[i]);
      if (mychild === b) {
        console.log(num + 1);
        process.exit();
      }
      if (!visit[mychild]) {
        visit[mychild] = true;
        find(mychild, num + 1);
      }
    }
  }

  const n = Number(input[iter++]);
  let command = input[iter++].split(' ');
  const a = Number(command[0]);
  const b = Number(command[1]);

  const parent = [...new Array(n + 1)].map(() => 0);
  const child = [...new Array(n + 1)].map(() => new Object());

  const m = Number(input[iter++]);
  for (let i = 0; i < m; i++) {
    command = input[iter++].split(' ');
    const x = Number(command[0]);
    const y = Number(command[1]);
    parent[y] = x;
    child[x][y] = true;
  }

  const visit = [...new Array(n + 1)].map(() => false);
  visit[a] = true;
  find(a, 0);
  console.log(-1);
  process.exit();
});
